docs/js-core-3/module-systems/lesson.md

```diff
@@ -0,0 +1,399 @@
+
+As we looked at earlier, we can also use file paths when importing files in JavaScript. We can now explain what this code does:
+
+```js
+const findWidget = require("./widget-finder");
+```
+
+Now we know that the `./widget-finder` file path refers to a file in the same folder called `widget-finder.js`. We can write out the `.js` bit if we want, but if you leave it off, then the module system will assume that you mean `.js`.
+
+So when we run this code, the module system will look for the `widget-finder.js` file and import it into the current file.
+
+### `module.exports`
+
+We've seen how we import code into our files, but we also need to see how we _export_ code too.
+
+To understand why we need to export as well as import, let's think about shipping products between countries in real-life. Some products we want to remain in the country, and some we want to export. With our code, we also want to choose what should only remain in the file, and what should be exported.
+
+Now, let's look at how we export code using the CommonJS module system:
+
+```js
+function findWidget(id) {
+  // ...
+}
+
+module.exports = findWidget;
+```
+
+Here, we've defined a `findWidget` function just like any other function. The important bit here is that we've then assigned a special `module.exports` variable to the `findWidget` reference. Remember that when we define a function like this, we also create a reference that can be used like any other variable.
+
+In this example, we're exporting only one "thing". This is often useful, but sometimes we may want to export multiple things. We can still do this with `module.exports` by instead exporting an object, which then contains the things we want to export:
+
+```js
+function findWidget() {
+  // ...
+}
+function findAllWidgets() {
+  // ...
+}
+
+module.exports = {
+  findWidget: findWidget,
+  findAllWidgets: findAllWidgets,
+};
+```
+
+Then when we import this file, we get access to the object that was exported, from which we can then get access to the things inside it:
+
+```js
+const widgetFinderExports = require("./widget-finder");
+
+const findWidget = widgetFinderExports.findWidget;
+const findAllWidgets = widgetFinderExports.findAllWidgets;
+```
+
+## NPM
+
+_Node Package Manager_ - or NPM for short - is a massive collection of JavaScript code that you can download and use in your apps (for free!). This is extremely helpful for us, as we can concentrate on the specific problem we want to solve, which helps build our apps faster.
+
+As the name implies, NPM is a tool for managing _packages_. So what is a package? Packages aren't anything special really! They're made up of JavaScript files, just like any other file that you've written. A developer has then uploaded those files into the NPM collection, allowing any one to dowload them and use them for free.
```

Comment:
> So what is a package

Doesn't really go on to answer the question. Perhaps:

> Packages aren't anything special really! They're made up of JavaScript files, just like any other file that you've written. **A a package is a collection of files that helps you to add some functionality to your application** A developer has then uploaded those files into the NPM collection, allowing any one to dowload them and use them for free.

taken from https://cleverbeagle.com/blog/articles/what-are-javascript-packages-and-dependencies